### PR TITLE
#8421: uiComponent sortOrder is not applied properly if some layout items are missing sortOrder value

### PR DESCRIPTION
**Pull request by RomaKis**
### Description
All ui components, which don't has SortOrder undefined will has SordOrder depends on as items are rendered as they appear in layout file

### Fixed Issues (if relevant)
1. magento/magento2#8421: Issue title

### Manual testing scenarios
1. Create a new uiComponent (or use existing one just for testing)
2. Try to add this to checkout and order after summary item in summary displayArea

```xml 
<referenceBlock name="checkout.root">
            <arguments>
                <argument name="jsLayout" xsi:type="array">
                    <item name="components" xsi:type="array">
                        <item name="checkout" xsi:type="array">
                            <item name="children" xsi:type="array">
                                <item name="sidebar" xsi:type="array">
                                    <item name="children" xsi:type="array">
                                        <item name="new_component" xsi:type="array">
                                            <item name="component" xsi:type="string">uiComponent</item>
                                            <item name="displayArea" xsi:type="string">summary</item>
                                            <item name="sortOrder" xsi:type="string">900</item>
                                            <item name="config" xsi:type="array">
                                                <item name="template" xsi:type="string">new_component/template</item>
                                            </item>
                                        </item>
                                    </item>
                                </item>
                            </item>
                        </item>
                    </item>
                </argument>
            </arguments>
        </referenceBlock>
```
3. This component must be placed as a last element as it has sortOrder 900.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/Ui/view/base/web/js/core/renderer/layout.js [hunk 1]
```diff
@@ -600,6 +602,25 @@ define([
...
             return result;
+        },
+
+        /**
+         * Get sort order for child if it is undefined.
+         *
+         * @param {Object} parent
+         * @param {Object} child
+         */
+        getChildSortOrderIfUndefined: function (parent, child) {
```
Comment (ishakhsuvarov) [inline app/code/Magento/Ui/view/base/web/js/core/renderer/layout.js]: This method does not return anything, but naming suggests so (`get`).

File: app/code/Magento/Ui/view/base/web/js/core/renderer/layout.js [hunk 2]
```diff
@@ -600,6 +602,25 @@ define([
...
+            var key, i = 0;
```
Comment (ishakhsuvarov) [inline app/code/Magento/Ui/view/base/web/js/core/renderer/layout.js]: While `i` is a common name for index in the loops, it is still suggested to use conscious names.

File: app/code/Magento/Ui/view/base/web/js/core/renderer/layout.js [hunk 3]
```diff
@@ -600,6 +602,25 @@ define([
...
+
+            for (key in parent.children) {
+                if (child.index === key
+                    && typeof parent.children[key].sortOrder === "undefined") {
```
Comment (ishakhsuvarov) [inline app/code/Magento/Ui/view/base/web/js/core/renderer/layout.js]: Looks like this component depends on `underscore` already, so it is possible to use it's method to detect undefined.

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]
Please attempt to update JS or integration tests as well to support this case.

File: app/code/Magento/Ui/view/base/web/js/core/renderer/layout.js (hunk 1 above)
Comment (RomaKis) [reply app/code/Magento/Ui/view/base/web/js/core/renderer/layout.js]: Fixed.

File: app/code/Magento/Ui/view/base/web/js/core/renderer/layout.js (hunk 2 above)
Comment (RomaKis) [reply app/code/Magento/Ui/view/base/web/js/core/renderer/layout.js]: Fixed.

File: app/code/Magento/Ui/view/base/web/js/core/renderer/layout.js (hunk 3 above)
Comment (RomaKis) [reply app/code/Magento/Ui/view/base/web/js/core/renderer/layout.js]: Fixed.

**Review by ishakhsuvarov — APPROVED** [review]

**Comment by omiroshnichenko** [comment]
Hi, @RomaKis, this pull request make no sense. I've described correct use of sortOrder in [issue comment](https://github.com/magento/magento2/issues/8421#issuecomment-342845954).

**Pull request closed without merging**